.github/workflows/autotools-build.yml

```diff
@@ -0,0 +1,54 @@
+name: ADFlib autotools
+
+on:
+  push:
+    branches: [ "citest" ]
+  pull_request:
+    branches: [ "citest" ]
+
+jobs:
+  ubuntu_native_default:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - name: install prereq.
+      run: sudo apt-get update ; sudo apt-get install -y check
+    - name: autogen
+      run: ./autogen.sh
+    - name: configure
+      run: ./configure
+    - name: make
+      run: make
+    - name: make check
+      run: make check
+    - name: make install
```

Comment:
It does not currently build a viable distribution package...

```
$ make distcheck
make[4]: *** No rule to make target 'doc_DATA', needed by 'distdir-am'.  Stop.
```

This is because of the line `EXTRA_DIST = doc_DATA nobase_doc_DATA` in doc/Makefile.am which misunderstands what EXTRA_DIST does - it's a list of _files_ that have not been mentioned in other variables but need to be included in the distribution tarball. There are no such files. Just remove this line.

Reply:
Moving on, if we do remove the line, we then get to:

```
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: error: cannot find input file: `src/generic/Makefile.in'
```

This happens because what would normally be included in the distribution by use of SUBDIRS= in src/Makefile.am is thwarted because it is a dynamic selection, and only one is selected. Distribution (i.e. building a tarball) needs all platforms' native source code! It needs to look more like his:

```
SUBDIRS = $(NATIVE_DIR)
EXTRA_DIST = generic linux win32 # because $(NATIVE_DIR) is dynamic
```

Reply:
Moving on, if we make that change we get to:

```
make[4]: Entering directory '/tmp/tw/adflib-0.7.13/_build/sub/regtests/Test'
make[4]: *** No rule to make target 'floppy.sh', needed by 'floppy.sh.log'.  Stop.
```

All the *.sh files in the regtests/Test/ directory seem to be needed to conduct the tests, but they're not referenced in Makefile.am so they won't be included in any distribution. There is probably a better solution, but this is one solution:

``` 
TESTS = floppy.sh bigdev.sh
EXTRA_DIST = floppy.sh bootdisk.sh del_test.sh dir_test.sh dir_test2.sh \
             dir_test_chdir.sh fl_test.sh fl_test2.sh file_test.sh \
             file_test2.sh file_test3.sh file_seek_test.sh file_seek_test2.sh \
             file_seek_after_write file_read_hard_link_test.sh rename.sh \
             undel.sh undel2.sh undel3.sh floppy_overfilling_test.sh \
             bigdev.sh hd_test.sh hd_test2.sh hd_test3.sh hardfile.sh
 ```

Reply:
Moving on, the next error is:

```
make[4]: Entering directory '/tmp/tw/adflib-0.7.13/_build/sub/tests'
PASS: test_adfDays2Date
PASS: test_adfPos2DataBlock
PASS: test_file_append
PASS: test_file_create
PASS: test_file_overwrite
PASS: test_file_overwrite2
PASS: test_file_seek_after_write
PASS: test_file_write
PASS: test_file_write_chunks
PASS: test_test_util
make[4]: *** No rule to make target 'test_examples', needed by 'test_examples.log'.  Stop.
```

Again, a potential fix:

```
check_SCRIPTS = test_examples
TESTS = $(check_PROGRAMS) $(check_SCRIPTS)
EXTRA_DIST = $(check_SCRIPTS) arccsh.adf
```

But this leads to failure of the test_example script, because it makes the assumption that srcdir = builddir (which is not always the case, and it's very helpful to packagers if that functionality works.

We can fix the script by asking it to look at srcdir, if that exists:

```bash
TEST_ADF=arccsh.adf
[ -n "$srcdir" ] && TEST_ADF="$srcdir/arccsh.adf"
... 
CMDS[0]="unadf -r $TEST_ADF"
...
```

Reply:
Here is the entire set of changes from above, and I'm also more than happy to apply this later if you want to say distcheck is broken for now and can be fixed later.

```diff
diff --git a/configure.ac b/configure.ac
index 1a7745c..757a0cf 100644
--- a/configure.ac
+++ b/configure.ac
@@ -98,9 +98,9 @@ AC_ARG_ENABLE([native_win32],
 echo "NATIVE_DIR: ${NATIVE_DIR}"
 echo "ADFLIB_OS: ${ADFLIB_OS}"
 
-#AM_CONDITIONAL([NATIVE_GENERIC], [test x${devnative} = xgeneric])
-#AM_CONDITIONAL([NATIVE_LINUX],   [test x${devnative} = xlinux])
-#AM_CONDITIONAL([NATIVE_WIN32],   [test x${devnative} = xwin32])
+AM_CONDITIONAL([NATIVE_GENERIC], [test x${NATIVE_DIR} = xgeneric])
+AM_CONDITIONAL([NATIVE_LINUX],   [test x${NATIVE_DIR} = xlinux])
+AM_CONDITIONAL([NATIVE_WIN32],   [test x${NATIVE_DIR} = xwin32])
 
 AM_CONDITIONAL([ADFLIB_OS_CYGWIN], [test x${ADFLIB_OS} = xcygwin])
 AM_CONDITIONAL([ADFLIB_OS_DARWIN], [test x${ADFLIB_OS} = xdarwin])
@@ -162,9 +162,6 @@ AC_SUBST(LT_AGE)
 
 AC_CONFIG_FILES([Makefile
                  src/Makefile
-                 src/generic/Makefile
-                 src/linux/Makefile
-                 src/win32/Makefile
                  doc/Makefile
                  examples/Makefile
                  regtests/Test/Makefile
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 40c04de..4b704af 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -20,5 +20,4 @@ nobase_doc_DATA = FAQ/image/adf_dir.gif \
                   API.txt \
                   version0.7.9d_gary.txt
 
-EXTRA_DIST = doc_DATA nobase_doc_DATA
-
+EXTRA_DIST = $(doc_DATA) $(nobase_doc_DATA)
diff --git a/examples/Makefile.am b/examples/Makefile.am
index c4e5a2a..6a75116 100644
--- a/examples/Makefile.am
+++ b/examples/Makefile.am
@@ -1,9 +1,6 @@
 bin_PROGRAMS = unadf adf_floppy_create adf_floppy_format adf_show_metadata
 
 ADFLIBS = $(top_builddir)/src/libadf.la
-if ADFLIB_OS_CYGWIN
-    ADFLIBS += $(top_builddir)/src/$(NATIVE_DIR)/libadfnative.a
-endif
 
 unadf_SOURCES = unadf.c
 unadf_LDADD = $(ADFLIBS)
diff --git a/regtests/Test/Makefile.am b/regtests/Test/Makefile.am
index 7bb4633..2dfb060 100644
--- a/regtests/Test/Makefile.am
+++ b/regtests/Test/Makefile.am
@@ -38,14 +38,18 @@ check_PROGRAMS = \
 	undel2 \
 	undel3
 
+check_SCRIPTS = floppy.sh bootdisk.sh del_test.sh dir_test.sh dir_test2.sh \
+                dir_test_chdir.sh fl_test.sh fl_test2.sh file_test.sh \
+                file_test2.sh file_test3.sh file_seek_test.sh file_seek_test2.sh \
+                file_seek_after_write file_read_hard_link_test.sh rename.sh \
+                undel.sh undel2.sh undel3.sh floppy_overfilling_test.sh \
+                bigdev.sh hd_test.sh hd_test2.sh hd_test3.sh hardfile.sh
+
 #TESTS = $(check_PROGRAMS)
 TESTS = floppy.sh bigdev.sh
-#EXTRA_DIST = $(TESTS)
+EXTRA_DIST = $(check_SCRIPTS)
 
 ADFLIBS = $(top_builddir)/src/libadf.la
-if ADFLIB_OS_CYGWIN
-    ADFLIBS += $(top_builddir)/src/$(NATIVE_DIR)/libadfnative.a
-endif
 
 fl_test_SOURCES = fl_test.c
 fl_test_LDADD = $(ADFLIBS)
diff --git a/src/Makefile.am b/src/Makefile.am
index 943b30f..8b5781b 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1,5 +1,4 @@
-
-SUBDIRS = $(NATIVE_DIR)
+AUTOMAKE_OPTIONS = subdir-objects
 
 lib_LTLIBRARIES = libadf.la
 
@@ -46,17 +45,17 @@ include_HEADERS = adf_defs.h \
 libadf_la_LDFLAGS = \
     -version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
 
-# Link adfnative to libadf
-if ADFLIB_OS_CYGWIN
-    # for some reason CygWin require full path(?)
-    libadf_la_LDFLAGS += -L$(abs_builddir)/$(NATIVE_DIR) -ladfnative
+if NATIVE_GENERIC
+libadf_la_SOURCES += generic/adf_nativ.c
+include_HEADERS += generic/adf_nativ.h
 endif
-if ADFLIB_OS_DARWIN
-    # for some reason Darwin require full path(?)
-    libadf_la_LDFLAGS += -L$(abs_builddir)/$(NATIVE_DIR) -ladfnative
+if NATIVE_LINUX
+libadf_la_SOURCES += linux/adf_nativ.c
+include_HEADERS += linux/adf_nativ.h
 endif
-if ADFLIB_OS_LINUX
-    libadf_la_LDFLAGS += -L$(NATIVE_DIR) -ladfnative
+if NATIVE_WIN32
+libadf_la_SOURCES += win32/adf_nativ.c win32/nt4_dev.c
+include_HEADERS += win32/adf_nativ.h win32/nt4_dev.h win32/defendian.h
 endif
 
 AM_CPPFLAGS = -D_XOPEN_SOURCE -D_SVID_SOURCE -D_BSD_SOURCE -D_DEFAULT_SOURCE \
diff --git a/src/generic/Makefile.am b/src/generic/Makefile.am
deleted file mode 100644
index 8a082d5..0000000
--- a/src/generic/Makefile.am
+++ /dev/null
@@ -1,19 +0,0 @@
-
-AM_CPPFLAGS = -D_XOPEN_SOURCE -D_SVID_SOURCE -D_BSD_SOURCE -D_DEFAULT_SOURCE \
-    -D_GNU_SOURCE -I$(topsrcdir)/src
-AM_CFLAGS = -std=c99 -pedantic -Wall -Wextra -fPIC \
-    -Werror-implicit-function-declaration \
-    -I$(top_srcdir)/src
-
-
-lib_LIBRARIES = libadfnative.a
-libadfnative_a_SOURCES = adf_nativ.c
-include_HEADERS = adf_nativ.h
-libadfnative_a_LIBADD = adf_nativ.o
-
-#lib_LTLIBRARIES = libadfnative.la
-#libadfnative_la_SOURCES = adf_nativ.c
-#include_HEADERS = adf_nativ.h
-##libadfnative_la_LDFLAGS = adf_nativ.o
-
-
diff --git a/src/linux/Makefile.am b/src/linux/Makefile.am
deleted file mode 100644
index 60b0097..0000000
--- a/src/linux/Makefile.am
+++ /dev/null
@@ -1,12 +0,0 @@
-
-AM_CPPFLAGS = -D_XOPEN_SOURCE -D_SVID_SOURCE -D_BSD_SOURCE -D_DEFAULT_SOURCE \
-    -D_GNU_SOURCE -I$(topsrcdir)/src
-AM_CFLAGS = -std=c99 -pedantic -Wall -Wextra -fPIC \
-    -Werror-implicit-function-declaration \
-    -I$(top_srcdir)/src
-
-
-lib_LIBRARIES = libadfnative.a
-libadfnative_a_SOURCES = adf_nativ.c
-include_HEADERS = adf_nativ.h
-libadfnative_a_LIBADD = adf_nativ.o
diff --git a/src/win32/Makefile.am b/src/win32/Makefile.am
deleted file mode 100644
index b1fd90e..0000000
--- a/src/win32/Makefile.am
+++ /dev/null
@@ -1,25 +0,0 @@
-
-AM_CPPFLAGS = -D_XOPEN_SOURCE -D_SVID_SOURCE -D_BSD_SOURCE -D_DEFAULT_SOURCE \
-    -D_GNU_SOURCE -I$(topsrcdir)/src
-AM_CFLAGS = -std=c99 -pedantic -Wall -Wextra -fPIC \
-    -Werror-implicit-function-declaration \
-    -I$(top_srcdir)/src
-
-lib_LIBRARIES = libadfnative.a
-
-libadfnative_a_SOURCES = adf_nativ.c \
-	nt4_dev.c
-
-include_HEADERS = adf_nativ.h \
-	defendian.h \
-	nt4_dev.h
-
-libadfnative_a_LIBADD = adf_nativ.o \
-	nt4_dev.o
-
-#lib_LTLIBRARIES = libadfnative.la
-#libadfnative_la_SOURCES = adf_nativ.c
-#include_HEADERS = adf_nativ.h
-##libadfnative_la_LDFLAGS = adf_nativ.o
-
-
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 2ef0f97..abff3d4 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -15,13 +15,11 @@ check_PROGRAMS = \
     test_file_write_chunks \
     test_test_util
 
-TESTS = $(check_PROGRAMS) \
-    test_examples
+check_SCRIPTS = test_examples
+TESTS = $(check_PROGRAMS) $(check_SCRIPTS)
+EXTRA_DIST = $(check_SCRIPTS) arccsh.adf
 
 ADFLIBS = $(top_builddir)/src/libadf.la
-if ADFLIB_OS_CYGWIN
-    ADFLIBS += $(top_builddir)/src/$(NATIVE_DIR)/libadfnative.a
-endif
 
 test_adfDays2Date_SOURCES = test_adfDays2Date.c
 test_adfDays2Date_CFLAGS = $(CHECK_CFLAGS)
diff --git a/tests/test_examples b/tests/test_examples
index 0c18e41..61df489 100755
--- a/tests/test_examples
+++ b/tests/test_examples
@@ -7,6 +7,9 @@ PATH=`pwd`/../examples:$PATH
 #echo "PATH: ${PATH}"
 #EXAMPLES_PATH=../examples
 
+TEST_ADF=arccsh.adf
+[ -n "$srcdir" ] && TEST_ADF="$srcdir/arccsh.adf"
+
 set -e
 
 run_cmd()
@@ -25,11 +28,11 @@ run_cmd()
     ${EXEC} $@
 }
 
-CMDS[0]="unadf -r arccsh.adf"
-CMDS[1]="adf_show_metadata arccsh.adf"
-CMDS[2]="adf_show_metadata arccsh.adf CSH"
-CMDS[3]="adf_show_metadata arccsh.adf c/"
-CMDS[4]="adf_show_metadata arccsh.adf l"
+CMDS[0]="unadf -r $TEST_ADF"
+CMDS[1]="adf_show_metadata $TEST_ADF"
+CMDS[2]="adf_show_metadata $TEST_ADF CSH"
+CMDS[3]="adf_show_metadata $TEST_ADF c/"
+CMDS[4]="adf_show_metadata $TEST_ADF l"
 CMDS[5]="adf_floppy_create testflopdd1.adf dd"
 CMDS[6]="adf_floppy_format testflopdd1.adf TestFlopDD1 1"
 CMDS[7]="adf_show_metadata testflopdd1.adf"
```

Reply:
Thanks a lot for this careful check and insightful info. Clearly the build system config, as the lib itself, require certain amount of changes if we want to improve it. Obviously, I am not able to overlook every detail for every case, so this helps a lot.

We can do exactly as you proposed, that is - we merge this, and then you can apply the corrections (we can also look into some unclear details, if any, sometimes it is not that obvious what is the best way; I did what worked for me but clearly it is not that way for everyone...).

Btw. since we have the small configuration for "actions", before doing the PRs I think we can ask people to make sure is passes before submitting a PR (unless special case, like a useful patch, but done by someone do not know how to update the build system to make it working), so that we do not have to check things clearly broken (as I had to do initially - with my own changes...).

And, of course, we can add `make distcheck` to the action's config (after we make it working with your fixes).